onnxruntime/contrib_ops/cuda/transformers/beam_search_topk.cu

```diff
@@ -0,0 +1,442 @@
+  typedef cub::BlockReduce<TopK<T, max_k>, thread_block_size> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+
+  input_values += vector_id * k * parts_per_beam;
+  input_tokens += vector_id * k * parts_per_beam;
+
+  TopK<T, max_k> thread_topk;
+  for (int i = 0; i < max_k; ++i) {
+    thread_topk.key[i] = -1;
+    thread_topk.value[i] = NumericLimits<T>::Min();
+  }
+
+  for (int idx = thread_id; idx < k * parts_per_beam; idx += thread_block_size) {
+    value_shared_buf[idx] = input_values[idx];
+    tokens_shared_buf[idx] = input_tokens[idx];
+  }
+  __syncthreads();
+
+  if (thread_id < parts_per_beam) {
+    T* b_v = value_shared_buf + thread_id * k;
+    int32_t* b_i = tokens_shared_buf + thread_id * k;
+    for (int i = 0; i < k; i++) {
+      thread_topk.Insert(b_v[i], b_i[i]);
+    }
+  }
+
+  TopK<T, max_k> topk_block = BlockReduce(temp_storage).Reduce(thread_topk, reduce_topk_op<T, max_k>);
+
+  if (thread_id == 0) {
+    output_values += vector_id * k;
+    output_indices += vector_id * k;
+
+    for (int i = 0; i < k; ++i) {
+      if (i < k) {
+        output_values[i] = topk_block.value[i];
+        output_indices[i] = topk_block.key[i];
+      }
+    }
+  }
+}
+
+template <typename T, int max_k>
+void LaunchBeamSearchOnlineTopKStage2Kernel(
+    const T* topk_values_tmp,
+    const int32_t* topk_indices_tmp,
+    int32_t batch_size,
+    int32_t num_beams,
+    int32_t vocab_size,
+    int32_t parts_per_beam,
+    int32_t K,
+    T* output_values,
+    int32_t* output_indices,
+    cudaStream_t stream) {
+  ORT_ENFORCE(parts_per_beam <= 128, "Parts per beam should not be greater than 128");
+
+  int smem_stage2_size = parts_per_beam * max_k * 2 * sizeof(int32_t);
+
+  if (parts_per_beam <= 32) {
+    BeamSearchOnlineTopKStage2Kernel<T, max_k, 32><<<batch_size * num_beams, 32, smem_stage2_size, stream>>>(
+        topk_values_tmp, topk_indices_tmp, K, vocab_size, parts_per_beam, output_values, output_indices);
+    return;
+  }
+
+  if (parts_per_beam <= 64) {
+    BeamSearchOnlineTopKStage2Kernel<T, max_k, 64><<<batch_size * num_beams, 64, smem_stage2_size, stream>>>(
+        topk_values_tmp, topk_indices_tmp, K, vocab_size, parts_per_beam, output_values, output_indices);
+    return;
+  }
+
+  BeamSearchOnlineTopKStage2Kernel<T, max_k, 128><<<batch_size * num_beams, 128, smem_stage2_size, stream>>>(
+      topk_values_tmp, topk_indices_tmp, K, vocab_size, parts_per_beam, output_values, output_indices);
+  return;
+}
+
+template <typename T, int max_k>
+void TopKLauncherMaxK(
+    const T* input,
+    int batch_size,
+    int num_beams,
+    int vocab_size,
+    int K,
+    T* output_values,
+    int32_t* output_indices,
+    T* output_values_tmp,
+    int32_t* output_indices_tmp,
+    cudaStream_t stream) {
+  constexpr int kThreadBlockSize = (max_k < 16) ? (max_k < 8) ? 256 : 128 : 64;
+
+  int voc_parts = 4;
+  if (batch_size * num_beams < 256) {
+    // volta has 80 SMs, so we aim for three waves
```

Comment:
Does this code work for older GPUs (sm < 70)?

Reply:
yes, it works but it runs more wave.